api/desecapi/tests/test_throttling_user.py

```diff
@@ -0,0 +1,80 @@
+from unittest import mock
+import time
+
+from django.core.cache import cache
+from django.test import TestCase, override_settings
+from rest_framework import status
+from rest_framework.response import Response
+from rest_framework.views import APIView
+from rest_framework.test import APIRequestFactory, force_authenticate
+
+from desecapi.models import User
+
+
+class MockView(APIView):
+    @property
+    def throttle_classes(self):
+        # Need to import here so that the module is only loaded once the settings override is in effect
+        from desecapi.throttling import UserRateThrottle
+
+        return (UserRateThrottle,)
+
+    def get(self, request):
+        return Response("foo")
+
+
+class ThrottlingTestCase(TestCase):
+    """
+    Based on DRF's test_throttling.py.
+    """
+
+    def setUp(self):
+        super().setUp()
+        self.factory = APIRequestFactory()
+
+    def _test_requests_are_throttled(self, counts, user=None):
+        def do_test():
+            view = MockView.as_view()
+            sum_delay = 0
+            for delay, count, max_wait in counts:
+                sum_delay += delay
+                with mock.patch(
+                    "desecapi.throttling.UserRateThrottle.timer",
+                    return_value=time.time() + sum_delay,
+                ):
+                    for _ in range(count):
+                        response = view(request)
+                        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+                    response = view(request)
+                    self.assertEqual(
+                        response.status_code, status.HTTP_429_TOO_MANY_REQUESTS
+                    )
+                    self.assertTrue(
+                        max_wait - 1 <= float(response["Retry-After"]) <= max_wait
+                    )
+
+        cache.clear()
+        request = self.factory.get("/")
+        if user is not None:
+            force_authenticate(request, user=user)
+        with override_settings(
+            REST_FRAMEWORK={"DEFAULT_THROTTLE_RATES": {"user": "10/d"}}
+        ):
+            do_test()
+
+    def test_requests_are_throttled_unauthenticated(self):
+        self._test_requests_are_throttled([(0, 10, 86400)])
```

Comment:
how can the 86400 requests be successful? I only see a limit of 10/d

Reply:
This is not the number of requests, but the expected "Retry-After" header value. :) The numbers say that it's trying 10 times, after which such a header with value 86400 is expected.

Reply:
Is the ability to run several tests with one call to `_test_requests_are_throttled` actually used anywhere? I see two calls, both only have a single entry in the `counts` array. I find that 'anonymous data structure' confusing, if it is not required it could be removed.

Reply:
That's an omission; I've now modified `test_requests_are_throttled_unauthenticated` to make multiple requests, in order to test that a new quota is made available after the wait time is over..